src/java_bytecode/java_types.cpp

```diff
@@ -233,14 +302,109 @@ typet java_type_from_string(const std::string &src)
   case 'Z': return java_boolean_type();
   case 'V': return java_void_type();
   case 'J': return java_long_type();
-
+  case 'T':
+  {
+    // parse name of type variable
+    INVARIANT(src[src.size()-1]==';', "Generic type name must end on ';'.");
+    PRECONDITION(!class_name.empty());
+    irep_idt type_var_name(class_name+"::"+src.substr(1, src.size()-2));
+    return java_generic_parametert(
+      type_var_name,
+      java_type_from_string("Ljava/lang/Object;").subtype());
+  }
   case 'L':
     {
       // ends on ;
       if(src[src.size()-1]!=';')
         return nil_typet();
       std::string class_name=src.substr(1, src.size()-2);
```

Comment:
as discussed, this shadows the `class_name` parameter and should be moved down to where this string is used (after the generic signature case)